### PR TITLE
DROOLS-1964: [DMN Editor] Input Data typeRef

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/DROOLS-1964

This also fixes [DROOLS-2561](https://issues.jboss.org/browse/DROOLS-2561) and [DROOLS-2562](https://issues.jboss.org/browse/DROOLS-2562)

### CODE REVIEW

File: kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/definition/v1_1/DMNModelInstrumentedBase.java [hunk 1]
```diff
@@ -53,4 +60,22 @@ public void setAdditionalAttributes(Map<QName, String> additionalAttributes) {
...
+        this.parent = parent;
+    }
+
+    public Optional<String> getPrefixForNamespaceURI(final String namespaceURI) {
+        if (this.nsContext != null && this.nsContext.containsValue(namespaceURI)) {
+            return this.nsContext.entrySet().stream().filter(kv -> kv.getValue().equals(namespaceURI)).findFirst().map(Map.Entry::getKey);
+        }
+        if (this.parent != null) {
+            return parent.getPrefixForNamespaceURI(namespaceURI);
+        }
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/definition/v1_1/DMNModelInstrumentedBase.java]: Updated (in commit to follow).

File: kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/definition/v1_1/InformationItem.java [hunk 2]
```diff
@@ -33,12 +38,19 @@
...
+    protected Name name;
 
-    @Property
-    @FormField(afterElement = "name")
     protected QName typeRef;
 
+    @Property
+    @FormField(afterElement = "name", type = QNameFieldType.class)
+    @Valid
+    protected QNameHolder typeRefHolder;
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/definition/v1_1/InformationItem.java]: Well.... the _real_ field is ```typeRef``` as used by the marshaller and other code; however in order to make it work with the Properties Panel (when a value is changed and Stunner attempts to update the object property) I needed to use this work-around of having ```QNameHolder```. Stunner looks for a property annotated ```@Property``` and this is only supported on classes not fields (so I could not annotate the field ```@Property QName typeRef```. See [here](https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/definition/annotation/Property.java#L35).

[the remark this answers is not in the document]
Comment (jomarko) [reply kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/definition/v1_1/InformationItem.java]: Thanks for an explanation.

File: kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/factory/AbstractDMNDiagramFactory.java [hunk 3]
```diff
@@ -61,4 +63,21 @@ private void updateProperties(final Node<Definition<DMNDiagram>, ?> diagramNode,
...
+                                              DMNModelInstrumentedBase.URI_FEEL);
+        }
+        if (!dmnDefinitions.getNsContext().containsValue(DMNModelInstrumentedBase.URI_DMN)) {
+            dmnDefinitions.getNsContext().put(DMNModelInstrumentedBase.PREFIX_DMN,
+                                              DMNModelInstrumentedBase.URI_DMN);
+        }
+        if (!dmnDefinitions.getNsContext().containsValue(DMNModelInstrumentedBase.URI_KIE)) {
+            dmnDefinitions.getNsContext().put(DMNModelInstrumentedBase.PREFIX_KIE,
+                                              DMNModelInstrumentedBase.URI_KIE);
+        }
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/factory/AbstractDMNDiagramFactory.java]: Great idea. I've done exactly as you proposed (and updated elsewhere affected by the change).

File: kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/property/dmn/QNamePropertyType.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        QNamePropertyType that = (QNamePropertyType) o;
+        if (!name.equals(that.name)) {
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        return name.hashCode();
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/property/dmn/QNamePropertyType.java]: Done.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/TypePickerWidgetTest.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+        picker.setDMNModel(dmnModel);
+
+        verify(qNameFieldConverter).setDMNModel(eq(dmnModel));
+        verify(typeSelector).clear();
+
+        final BuiltInType[] bits = BuiltInType.values();
+        verify(picker, times(bits.length)).makeTypeSelector(builtInTypeCaptor.capture());
+
+        final List<BuiltInType> builtInTypes = new ArrayList<>(Arrays.asList(bits));
+        builtInTypes.removeAll(builtInTypeCaptor.getAllValues());
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/TypePickerWidgetTest.java]: Done.

**Comment by manstis** [comment]
@jomarko All updates applied in fourth commit.

**Pull request merged**